metricflow/model/semantics/join_validator.py

```diff
@@ -58,34 +75,140 @@ class DataSourceJoinValidator:
+                left_data_source_reference=left_data_source_reference,
+                known_data_source_joins=data_source_joins,
+                identifier_reference=identifier_reference,
+            )
+            for (joinable_data_source_reference, join_type) in joinable_data_source_references:
+                data_source_joins[joinable_data_source_reference.data_source_name] = DataSourceJoin(
+                    left_data_source_reference=left_data_source_reference,
+                    right_data_source_reference=joinable_data_source_reference,
+                    identifier_joins=[
+                        DataSourceIdentifierJoin(identifier_reference=identifier_reference, join_type=join_type)
+                    ],
+                )
+
+                if not include_multi_hop:
+                    continue
+
+                # Verify multi-hop joins
+                joinable_data_source = self._data_source_semantics.get_by_reference(
+                    data_source_reference=joinable_data_source_reference
+                )
+                assert joinable_data_source is not None
+                for secondary_identifier in joinable_data_source.identifiers:
+                    secondary_identifier_reference = IdentifierReference(element_name=secondary_identifier.name)
+                    secondary_joinable_data_source_references = self._get_new_joinable_data_sources_for_identifier(
+                        left_data_source_reference=joinable_data_source_reference,
+                        known_data_source_joins=data_source_joins,
+                        identifier_reference=secondary_identifier_reference,
+                    )
+                    for (
+                        secondary_joinable_data_source_reference,
+                        secondary_join_type,
+                    ) in secondary_joinable_data_source_references:
+                        if (
+                            secondary_joinable_data_source_reference.data_source_name
+                            == left_data_source_reference.data_source_name
+                        ):
+                            continue
+                        data_source_joins[secondary_joinable_data_source_reference.data_source_name] = DataSourceJoin(
+                            left_data_source_reference=left_data_source_reference,
+                            right_data_source_reference=secondary_joinable_data_source_reference,
+                            identifier_joins=[
+                                DataSourceIdentifierJoin(
+                                    identifier_reference=identifier_reference, join_type=join_type
+                                ),
+                                DataSourceIdentifierJoin(
+                                    identifier_reference=secondary_identifier_reference, join_type=secondary_join_type
+                                ),
+                            ],
+                        )
```

Comment:
This nested logic is quite confusing for me. This might be a case where operating on batches will be easier to understand, at least at the top level of this function. Something like:

```
# we can wrap this in a function and call it per data source, passing in an optional memoize list to cut out repeated work 
data_sources = {
    identifier.reference: self._data_source_semantics.get_data_sources_for_identifier(identifier.reference)
    for identifier in data_source.identifiers
}
valid_targets = {
    idref: filter(candidate_sources, lambda x: self.is_valid_data_source_join(...)) 
    for idref: candidate_sources in data_sources
}
# This produces a data_source_reference -> List[DataSourceIdentifierJoin]
return convert_to_data_source_to_join_path(valid_targets)
```

We'd have to do some extra stuff for making sure the multi-hop calls include all leading identifier references.

Let me know what you think! Note I like these kind of pipeline processing flows, so I might be missing an easier readability improvement.

Reply:
Yeah I think I will need to refactor this to make it recursive anyway since we want to support more than 2 hops! So l'll see if this structure will work for that while I refactor.